overlord/fdestate/fdestate.go

```diff
@@ -403,3 +404,106 @@ func MockSecbootGetPCRHandle(f func(devicePath, keySlot, keyFile string, hintExp
 		secbootGetPCRHandle = old
 	}
 }
+
+// KeyslotTarget uniquely identifies a target key slot by
+// its container role and name.
+type KeyslotTarget struct {
+	ContainerRole string `json:"container-role"`
+	Name          string `json:"name"`
+}
+
+func (k KeyslotTarget) String() string {
+	return fmt.Sprintf("(container-role: %q, name: %q)", k.ContainerRole, k.Name)
+}
+
+func (k KeyslotTarget) Validate() error {
+	if len(k.ContainerRole) == 0 {
+		return errors.New("key slot container role cannot be empty")
+	}
+	if k.ContainerRole != "system-data" && k.ContainerRole != "system-save" {
+		return fmt.Errorf(`unsupported key slot container role %q, expected "system-data" or "system-save"`, k.ContainerRole)
+	}
+
+	if len(k.Name) == 0 {
+		return errors.New("key slot name cannot be empty")
+	}
+	return nil
+}
+
+func checkRecoveryKeyIDExists(st *state.State, recoveryKeyID string) error {
+	mgr := fdeMgr(st)
+	rkeyInfo, err := mgr.recoveryKeyCache.Key(recoveryKeyID)
+	if err != nil {
+		return err
+	}
+	if rkeyInfo.Expired(time.Now()) {
+		return errors.New("recovery key has expired")
+	}
+	return nil
+}
+
+const tmpKeyslotPrefix = "snapd-tmp"
+
+func tmpKeyslotTarget(keyslot KeyslotTarget) KeyslotTarget {
+	return KeyslotTarget{
+		ContainerRole: keyslot.ContainerRole,
+		Name:          fmt.Sprintf("%s:%s", tmpKeyslotPrefix, keyslot.Name),
+	}
+}
+
+// ReplaceRecoveryKey creates a taskset that replaces the
+// recovery key for the specified target key slots using
+// the recovery key identified by recoveryKeyID.
+//
+// If keyslots is empty, the "default-recovery" key slot is
+// used by default for both the "system-data" and "system-save"
+// container roles.
+func ReplaceRecoveryKey(st *state.State, recoveryKeyID string, keyslots []KeyslotTarget) (*state.TaskSet, error) {
+	if len(keyslots) == 0 {
+		// target default-recovery key slots by default if no key slot targets are specified
+		keyslots = append(keyslots,
+			KeyslotTarget{ContainerRole: "system-data", Name: "default-recovery"},
+			KeyslotTarget{ContainerRole: "system-save", Name: "default-recovery"},
+		)
+	}
+
+	tmpKeyslots := make([]KeyslotTarget, 0, len(keyslots))
+	tmpKeyslotRenames := make([]string, 0, len(keyslots))
+	for _, keyslot := range keyslots {
+		if err := keyslot.Validate(); err != nil {
```

Comment:
I think `keyslot.Validate` here is a bit confusing. Reading the code for `Validate`, it does not look like it is validating a keyslot. But the identification of a keyslot.

"Validating a keyslot" sounds more like validating the key data of the key slot.

Not sure what to do however here to make it less confusing.